### PR TITLE
EKF2: Error-State Kalman Filter

### PR DESCRIPTION
**Pull request by bresch**
### Solution
Error-state EKF based on [Quaternion kinematics for the error-state Kalman filter](https://arxiv.org/pdf/1711.02508.pdf)

Main advantages of the "error state" derivation compared to the current "total state":
- better convergence properties (the error of the state behaves much more like a linear system than the state itself)
- more stable numerically (less equations in general)
- less computationally expensive (see comparison on real hardware below)
- correct way of describing the quaternion uncertainty using Lie group theory (the uncertainty is defined like the error quaternion: on the tangent space and not as a 4D vector). Note that because of this, the state vector and state covariance have different sizes.

Disadvantages:
- the derivation is a bit more difficult... I'm not aware of any other drawbacks

Flash usage on FMU v5 (saves 8.3kB of flash):
```
FLASH_AXIM:     2037485 B      2016 KB     98.70%
```
compared to main:
```
FLASH_AXIM:     2045789 B      2016 KB     99.10%
```

CPU usage (CPU load reduction of ~0.3%, less stack usage):
```
ekf2: ECL full update: 4108 events, 860456us elapsed, 209.46us avg, min 135us max 606us 81.220us rms
ekf2: ECL update: 4109 events, 2865us elapsed, 0.69us avg, min 0us max 48us 1.099us rms

 PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) STATE FD
1254 wq:INS0                        16  8.262  3652/ 5976 241 (241)  w:sem  4
```
compared to main:
```
ekf2: ECL full update: 5006 events, 1163229us elapsed, 232.37us avg, min 152us max 682us 85.309us rms
ekf2: ECL update: 5006 events, 3584us elapsed, 0.71us avg, min 0us max 49us 1.376us rms

 PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) STATE FD
1254 wq:INS0                        17  8.625  4212/ 5976 241 (241)  w:sem  4
```

### Changelog Entry
For release notes:
```
Change EKF2 derivation to an error state formulation (ESKF)
New parameter: -
Documentation: TBD
```

### Test coverage
- unit tests
- sitl tests (see TODO below)
- replay on log with optical flow

TODO:
- [x] fix `--disable_mag` and `--disable_wind` options in `derivation.py`
- [x] test mc wind estimation https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1782809616
- [x] test heading convergence without mag https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1782836331
- [x] test gyro bias estimation https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1782860712
- [x] test accel bias estimation https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1782868804
- [x] test mag bias estimation https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1782819797
- [x] test airspeed fusion https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1782905921
- [x] fast yaw spin for > 1min https://github.com/PX4/PX4-Autopilot/pull/22262#issuecomment-1784759011
- [x] real test flights
    - Indoor optical flow (stab, alt and pos modes): https://logs.px4.io/plot_app?log=d533e649-72e6-416c-a8a0-b0e0334cf62f no issues
    - VTOL plane, hover and FW flight: https://logs.px4.io/plot_app?log=2a00726f-cd5b-4a4e-82e7-f6e1ee9d8b80
    - Wrong initial heading: https://logs.px4.io/plot_app?log=c6e7094b-7213-488b-8eb0-5ea7b4b9b77f
    - VTOL plane mission flight: https://logs.px4.io/plot_app?log=eece1b94-3eed-4f6a-a1c9-1dbc9ded0679

### CODE REVIEW

File: src/modules/ekf2/EKF/ekf.h [hunk 1]
```diff
@@ -234,7 +232,7 @@ class Ekf final : public EstimatorInterface
 #endif // CONFIG_EKF2_GRAVITY_FUSION
 
 	// get the state vector at the delayed time horizon
-	const matrix::Vector<float, State::size> &getStateAtFusionHorizonAsVector() const { return _state.vector(); }
+	const matrix::Vector<float, 24> &getStateAtFusionHorizonAsVector() const { return _state.vector(); }
```
Comment (bresch) [inline src/modules/ekf2/EKF/ekf.h:235]: need to fix that hardcoded 24

Comment (dagar) [reply src/modules/ekf2/EKF/ekf.h]: We also have https://github.com/PX4/PX4-Autopilot/blob/a2aa2ff457fe1d4268ede49348a6938bc0917dbe/src/modules/ekf2/EKF/ekf.h#L87

Comment (bresch) [reply src/modules/ekf2/EKF/ekf.h]: done

**Comment by bresch** [comment]
Mag x and y offset parameters set to 0.2
Screenshot from 2023-10-27 14-18-01

**Comment by bresch** [comment]
Heading convergence from a 90 degrees error:
Screenshot from 2023-10-27 14-27-14

Compared to the same conditions with the previous derivation (note that at t=100 the heading got automatically reset to the yaw emergency estimator)
Screenshot from 2023-10-27 14-27-46

**Comment by bresch** [comment]
With different gyro biases (set in the cal parameters as offsets. X is set on ground, Y and Z are set in air):
Screenshot from 2023-10-27 14-46-53

**Comment by bresch** [comment]
With different accel biases (initial x axis set on ground, y and z in air):
Screenshot from 2023-10-27 14-52-35

**Comment by bresch** [comment]
Airspeed fusion:
This isn't a great test as there is no wind in simulation, but we can see that the estimate seems to make sense
Screenshot from 2023-10-27 15-18-09
The actual accuracy should be tested in a real flight test later, but I don't expect any issue with this as the derivation didn't really change anything for the wind states.

**Comment by bresch** [comment]
400 deg/s yaw spin for over a minute. Also combined with some roll and pitch motions at the end. No visible deviation from the groundtruth.
Screenshot from 2023-10-30 10-02-41

**Comment by AlexKlimaj** [comment]
ARK Jetson, ARKV6X, ARK GPS, ARK Flow
https://review.px4.io/plot_app?log=4cd6e706-4e15-42ee-aedd-95f5792fa944

**Comment by bresch** [comment]
Thanks @AlexKlimaj for the log. I think it looks all good (except the height being pulled by the drifting GNSS altitude, but that's another topic :D).

**Comment by AlexKlimaj** [comment]
Flew under the deck which caused the GNSS fusion to stop then flew back out.
https://review.px4.io/plot_app?log=d6be638f-4ca1-4ed9-ae6b-8969ea724a9a

**Comment by AlexKlimaj** [comment]
650 with Flow and moving base heading.
https://review.px4.io/plot_app?log=dd1198a5-d0ef-48ce-a361-ca0dfe8edd9c

**Comment by jkflying** [comment]
> Heading convergence from a 90 degrees error: Screenshot from 2023-10-27 14-27-14
> 
> Compared to the same conditions with the previous derivation (note that at t=100 the heading got automatically reset to the yaw emergency estimator) Screenshot from 2023-10-27 14-27-46

Really cool PR.

Just curious, it looks like the old formulation (eventually) converges back to a closer heading than the error-state formulation. Could this be caused by some other biases etc. being updated? Maybe the innovation gates for these need to be retuned?

**Comment by bresch** [comment]
Thanks!

It converges closer to the true heading because mainly because there is more horizontal motion (longer heading observability period). In the test with the new derivation, the drone stabilizes much faster so the heading estimate cannot continue to converge as there is no motion anymore.

**Pull request merged**